File: app/src/main/java/com/siju/acexplorer/model/helper/FileOperations.java
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.model.helper;

import android.annotation.TargetApi;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.v4.provider.DocumentFile;
import android.util.Log;

import com.siju.acexplorer.AceApplication;
import com.siju.acexplorer.logging.Logger;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.channels.FileChannel;

import static com.siju.acexplorer.model.StorageUtils.getDocumentFile;
import static com.siju.acexplorer.model.StorageUtils.isOnExtSdCard;
import static com.siju.acexplorer.model.helper.SdkHelper.isAtleastLollipop;
import static com.siju.acexplorer.model.helper.SdkHelper.isKitkat;


public class FileOperations {

    private static final String TAG = "FileOperations";


    /**
     * Rename a folder. In case of extSdCard in Kitkat, the old folder stays in place, but files are moved.
     *
     * @param source The source folder.
     * @param target The target folder.
     * @return true if the renaming was successful.
     */
    @TargetApi(Build.VERSION_CODES.KITKAT)
    public static boolean renameFolder(@NonNull final File source, @NonNull final File target) {
        // First try the normal rename.
        if (rename(source, target.getName())) {
            return true;
        }
        if (target.exists()) {
            return false;
        }

        // Try the Storage Access Framework if it is just a rename within the same parent folder.
        if (isAtleastLollipop()
                && source.getParent().equals(target.getParent()) && isOnExtSdCard(source)) {

            DocumentFile document = getDocumentFile(source, true);

            Log.d(TAG, " Document uri in rename=" + document);
            if (document != null && document.renameTo(target.getName())) {
                return true;
            }
        }

        if (source.isFile()) {
            Logger.log(TAG, "Rename--root=");


            if (!mkfile(target)) {
                return false;
            }

        } else {
            // Try the manual way, moving files individually.
            if (!mkdir(target)) {
                return false;
            }
        }

        File[] sourceFiles = source.listFiles();

        if (sourceFiles == null || sourceFiles.length == 0) {
            return true;
        }

        for (File sourceFile : sourceFiles) {
            String fileName = sourceFile.getName();
            File targetFile = new File(target, fileName);
            if (!copyFile(sourceFile, targetFile)) {
                // stop on first error
                return false;
            }
        }
        // Only after successfully copying all files, delete files on source folder.
        for (File sourceFile : sourceFiles) {
            if (!FileUtils.deleteFile(sourceFile)) {
                // stop on first error
                return false;
            }
        }
        return true;
    }

    public static boolean mkfile(final File file) {
        if (file == null)
            return false;
        if (file.exists()) {
            // nothing to create.
            return !file.isDirectory();
        }

        // Try the normal way
        try {
            if (file.createNewFile()) {
                return true;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        // Even after exception, In Kitkat file might have got created so return true
        if (file.exists()) return true;
        boolean result;
        // Try with Storage Access Framework.
        if (isAtleastLollipop() && isOnExtSdCard(file)) {
            DocumentFile document = getDocumentFile(file.getParentFile(), true);
            // getDocumentFile implicitly creates the directory.
            try {
                Logger.log(TAG, "mkfile--doc=" + document);
                result = document != null && document.createFile(FileUtils.getMimeType(file), file.getName()) != null;
                return result;
            } catch (Exception e) {
                e.printStackTrace();
                return false;
            }
        }

        if (isKitkat()) {
            try {
                return MediaStoreHack.mkfile(file);
            } catch (Exception e) {
                return false;
            }
        }
        return false;
    }

    /**
     * Create a folder. The folder may even be on external SD card for Kitkat.
     *
     * @param file The folder to be created.
     * @return True if creation was successful.
     */
    public static boolean mkdir(final File file) {
        if (file == null)
            return false;
        if (file.exists()) {
            // nothing to create.
            return file.isDirectory();
        }

        // Try the normal way
        if (file.mkdirs()) {
            return true;
        }

        // Try with Storage Access Framework.
        if (isAtleastLollipop() && isOnExtSdCard(file)) {
            DocumentFile document = getDocumentFile(file, true);
            // getDocumentFile implicitly creates the directory.
            return document != null && document.exists();
        }

        // Try the Kitkat workaround.
        if (isKitkat()) {
            try {
                return MediaStoreHack.mkdir(file);
            } catch (IOException e) {
                return false;
            }
        }

        return false;
    }


    /**
     * Copy a file. The target file may even be on external SD card for Kitkat.
     *
     * @param source The source file
     * @param target The target file
     * @return true if the copying was successful.
     */
    @SuppressWarnings("null")
    private static boolean copyFile(final File source, final File target) {
        FileInputStream inStream = null;
        OutputStream outStream = null;
        FileChannel inChannel = null;
        FileChannel outChannel = null;
        try {
            inStream = new FileInputStream(source);

            // First try the normal way
            if (FileUtils.isWritable(target)) {
                // standard way
                outStream = new FileOutputStream(target);
                inChannel = inStream.getChannel();
                outChannel = ((FileOutputStream) outStream).getChannel();
                inChannel.transferTo(0, inChannel.size(), outChannel);
            } else {
                if (isAtleastLollipop()) {
                    // Storage Access Framework
                    DocumentFile targetDocument = getDocumentFile(target, false);
                    if (targetDocument != null)
                        outStream =
                                AceApplication.getAppContext().getContentResolver().openOutputStream(targetDocument.getUri());
                } else if (isKitkat()) {
                    // Workaround for Kitkat ext SD card
                    Uri uri = MediaStoreHack.getUriFromFile(target.getAbsolutePath());
                    if (uri != null)
                        outStream = AceApplication.getAppContext().getContentResolver().openOutputStream(uri);
                } else {
                    return false;
                }

                if (outStream != null) {
                    // Both for SAF and for Kitkat, write to output stream.
                    byte[] buffer = new byte[16384]; // MAGIC_NUMBER
                    int bytesRead;
                    while ((bytesRead = inStream.read(buffer)) != -1) {
                        outStream.write(buffer, 0, bytesRead);
                    }
                }

            }
        } catch (Exception e) {
            Log.e(TAG,
                    "Error when copying file from " + source.getAbsolutePath() + " to " + target.getAbsolutePath(), e);
            return false;
        } finally {
            try {
                if (inStream != null) inStream.close();
                if (outStream != null) outStream.close();
                if (inChannel != null) inChannel.close();
                if (outChannel != null) outChannel.close();

            } catch (IOException e) {
                e.printStackTrace();
            }

        }
        return true;
    }

    private static boolean rename(File f, String name) {
        String newname = f.getParent() + "/" + name;
        if (f.getParentFile().canWrite()) {
            Logger.log(TAG, "Rename--canWrite=" + true);
            return f.renameTo(new File(newname));
        }
        return false;
    }


}

File: app/src/main/java/com/siju/acexplorer/main/model/helper/FileOperations.kt
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.main.model.helper

import android.util.Log
import com.siju.acexplorer.AceApplication
import com.siju.acexplorer.extensions.getMimeType
import com.siju.acexplorer.logging.Logger
import com.siju.acexplorer.main.model.StorageUtils.getDocumentFile
import com.siju.acexplorer.main.model.StorageUtils.isOnExtSdCard
import com.siju.acexplorer.storage.model.operations.DeleteOperation
import java.io.*
import java.nio.channels.FileChannel


private const val TAG = "FileOperations"
private const val BUFFER_SIZE_KB = 16384

object FileOperations {

    fun renameFile(source: File, target: File): Boolean {
        // First try the normal rename.
        if (rename(source, target)) {
            return true
        }
        if (target.exists()) {
            return false
        }

        // Try the Storage Access Framework if it is just a rename within the same parent folder.
        if (source.parent == target.parent && isOnExtSdCard(source)) {

            val document = getDocumentFile(source, true)

            Logger.log(TAG, " Document uri in rename=$document")
            if (document?.renameTo(target.name) == true) {
                return true
            }
        }

        if (source.isFile) {
            Logger.log(TAG, "Rename--root=")

            if (!mkfile(target)) {
                return false
            }

        }
        else {
            // Try the manual way, moving files individually.
            if (!mkdir(target)) {
                return false
            }
        }

        val sourceFileList = source.listFiles()

        if (sourceFileList.isNullOrEmpty()) {
            return true
        }

        val filesToDelete = arrayListOf<String>()
        for (sourceFile in sourceFileList) {
            val fileName = sourceFile.name
            val targetFile = File(target, fileName)
            if (!copyFile(sourceFile, targetFile)) {
                // stop on first error
                return false
            }
            filesToDelete.add(sourceFile.absolutePath)
        }
        val deleteOperation = DeleteOperation()
        // Only after successfully copying all files, delete files on source folder.
        for (sourceFile in sourceFileList) {
            if (!deleteOperation.delete(sourceFile)) {
                // stop on first error
                return false
            }
        }
        return true
    }

    fun mkfile(file: File?): Boolean {
        if (file == null)
            return false
        if (file.exists()) {
            // nothing to create.
            return !file.isDirectory
        }

        // Try the normal way
        try {
            return file.createNewFile()
        }
        catch (e: IOException) {
            e.printStackTrace()
        }

        if (file.exists()) {
            return true
        }
        val result: Boolean
        // Try with Storage Access Framework.
        if (isOnExtSdCard(file)) {
            val document = getDocumentFile(file.parentFile, true)
            // getDocumentFile implicitly creates the directory.
            return try {
                Logger.log(TAG, "mkfile--doc=$document")
                result = document?.createFile(file.getMimeType(),
                                              file.name) != null
                result
            }
            catch (e: Exception) {
                e.printStackTrace()
                false
            }

        }
        return false
    }

    fun mkdir(file: File?): Boolean {
        if (file == null)
            return false
        if (file.exists()) {
            // nothing to create.
            return file.isDirectory
        }

        // Try the normal way
        if (file.mkdirs()) {
            return true
        }

        // Try with Storage Access Framework.
        if (isOnExtSdCard(file)) {
            val document = getDocumentFile(file, true)
            // getDocumentFile implicitly creates the directory.
            return document != null && document.exists()
        }
        return false
    }


    private fun copyFile(source: File, target: File): Boolean {
        var inputStream: FileInputStream? = null
        var outputStream: OutputStream? = null
        var inChannel: FileChannel? = null
        var outChannel: FileChannel? = null
        try {
            inputStream = FileInputStream(source)

            // First try the normal way
            if (FileUtils.isWritable(target)) {
                // standard way
                outputStream = FileOutputStream(target)
                inChannel = inputStream.channel
                outChannel = outputStream.channel
                inChannel?.transferTo(0, inChannel.size(), outChannel)
            }
            else {
                // Storage Access Framework
                val targetDocument = getDocumentFile(target, false)
                if (targetDocument != null)
                    outputStream = AceApplication.appContext.contentResolver.openOutputStream(
                            targetDocument.uri)

                outputStream?.let {
                    // For SAF, write to output stream.
                    val buffer = ByteArray(BUFFER_SIZE_KB)
                    var bytesRead = inputStream.read(buffer)

                    while (bytesRead != -1) {
                        outputStream.write(buffer, 0, bytesRead)
                        bytesRead = inputStream.read(buffer)
                    }
                }

            }
        }
        catch (e: Exception) {
            Log.e(TAG,
                  "Error when copying file from " + source.absolutePath + " to " + target.absolutePath,
                  e)
            return false
        }
        finally {
            try {
                inputStream?.close()
                outputStream?.close()
                inChannel?.close()
                outChannel?.close()

            }
            catch (e: IOException) {
                e.printStackTrace()
            }

        }
        return true
    }

    private fun rename(file: File, newFile: File): Boolean {
        val parentFile = file.parentFile
        parentFile ?: return false
        if (parentFile.canWrite()) {
            Logger.log(TAG, "file:$file, newFIle:$newFile")
            return try {
                file.renameTo(newFile)
            } catch (exception : SecurityException) {
                false
            } catch (exception : NullPointerException) {
                false
            }
        }
        return false
    }

}
